src/Microsoft.AspNetCore.Mvc.Core/ControllerAttribute.cs

```diff
@@ -0,0 +1,17 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.AspNetCore.Mvc
+{
+    /// <summary>
+    /// Indicates that the type and any derived types that this attribute is applied to
+    /// are considered a controller by the default controller discovery mechanism, unless
+    /// <see cref="NonControllerAttribute"/> is applied to any type in the hierarchy.
```

Comment:
Does this mean any type with both `[Controller]` and `[NonController]` applied to it or its base classes will _not_ be a controller? I'm wondering in particular about the case where a base class has `[NonController]` and the type itself (or any closer base class) has `[Controller]`. Users may expect the nearest attribute overrides the other(s).

One alternative would be to remove `[NonController]` and then add a `bool IsController` property to this type. Get the `[Controller]` (inherited or not) and its `IsController` value wins. That would also avoid the compiler allowing odd cases like `[Controller]` and `[NonController]` on the _same_ type.